client_test.go

```diff
@@ -128,3 +134,223 @@ func TestNewClientWithFileConfig(t *testing.T) {
+//
+// The desired behavior is that a User-Agent header is included in the requests
+// when the supplied value > 0 or omitted from HTTP requests when the supplied
+// value length is 0.
+func (o *userAgentChecker) RoundTrip(req *http.Request) (*http.Response, error) {
+	reqBytes, err := httputil.DumpRequest(req, true)
+	require.NoError(o.t, err)
+	o.t.Log(string(reqBytes))
+
+	if o.expectedUserAgent == "" {
+		assert.NotContains(o.t, req.Header, headers.UserAgent,
+			"User-Agent header should be omitted when the supplied userAgent length is 0")
+	} else {
+		assert.Equal(o.t, o.expectedUserAgent, req.Header.Get(headers.UserAgent),
+			"User-Agent header should be included when the supplied userAgent length is >0")
+	}
+
+	return nil, o.expectedError
+}
+
+// NewHTTPClient returns an http.Client with this test helper as its transport
+// (see RoundTripper).
+func (o *userAgentChecker) NewHTTPClient() *http.Client {
+	return &http.Client{
+		Transport: o,
+	}
+}
+
+// RequireRoundTripError is used to check the error returned by any function
+// that has been instrumented with the NewHTTPClient (above). The instrumented
+// function is therefore expected to generate HTTP requests and where it does,
+// it is expected to return or wrap the error that is always returned by the
+// RoundTripper of this test helper.
+func (o *userAgentChecker) RequireRoundTripError(err error) {
+	require.ErrorContains(o.t, err, o.expectedError.Error(),
+		"The user supplied HTTP client (with simulated-error RoundTripper) should always be used, "+
+			"so the simulated-error should always be returned here. "+
+			"If not, it indicates one of two programming errors: "+
+			"1. the function is calling API endpoints with the wrong HTTP client, or "+
+			"2. the function is ignoring or hiding the error returned in the HTTP response.")
+}
+
+// TestNewClient_Connectors checks that all connectors are consistent in the way
+// they add or omit the User-Agent header.
+func TestNewClient_UserAgent(t *testing.T) {
+	// These base connector configs will be tested
+	connectorConfigs := []Config{
+		{
+			ConnectorType: endpoint.ConnectorTypeCloud,
+		},
+		{
+			ConnectorType: endpoint.ConnectorTypeTPP,
+			BaseUrl:       "https://tpp.example.local",
+		},
+		{
+			ConnectorType: endpoint.ConnectorTypeFirefly,
+			BaseUrl:       "https://firefly.example.local",
+		},
+	}
+
+	// The Authenticate method will be called with these connector specific
+	// credentials, which are chosen to trigger a request to an Auth endpoint.
+	connectorCredentials := map[endpoint.ConnectorType]endpoint.Authentication{
+		endpoint.ConnectorTypeCloud: endpoint.Authentication{
+			APIKey: "fake-key",
+		},
+		endpoint.ConnectorTypeTPP: endpoint.Authentication{
+			User:     "fake-user",
+			Password: "fake-password",
+		},
+		endpoint.ConnectorTypeFirefly: endpoint.Authentication{
+			IdentityProvider: &endpoint.OAuthProvider{
+				DeviceURL: "https://device.oauth.example.local",
+			},
+		},
```

Comment:
If you change this to use  `TokenURL` you find that the Firefly connector is not using the vcert.Client http Client, but is using calling an oauth2 function which uses the `http.DefaultClient`.

So the requests to the TokenURL will not have the standard User-Agent.
* https://github.com/golang/oauth2/issues/179

There is a way to set it, by adding our http.Client to the context and then adding a RoundTripper wrapper which injects the User-Agent, but I propose postponing that for another PR.

Reply:
Agreed. There's no analytics on Firefly at the moment that would take advantage of this yet anyway, but maybe something we can ask @tr1ck3r to determine if there's value there.  Agree to leave it to another PR after we reconcile that part.